File: examples/log_analyzer.nim
import std/[re, tables]

# OpenClaw log analyzer (Nim) — parses gateway access logs from stdin
let pattern = re"^(\S+)\s+(INFO|WARN|ERROR)\s+(\S+)\s+(.+)$"
var counts = {"INFO": 0, "WARN": 0, "ERROR": 0}.toTable

for line in stdin.lines:
  var m: array[4, string]
  if line.match(pattern, m):
    let level = m[1]
    counts[level] += 1
    if level == "ERROR":
      echo "⚠ ", m[0], " [", m[2], "] ", m[3]

echo "\n--- Summary ---"
for level in ["ERROR", "INFO", "WARN"]:
  echo level, ": ", counts[level]

File: examples/log-analyzer.rs
use std::collections::BTreeMap;
use std::io::{self, BufRead};

// OpenClaw log analyzer (Rust) — parses gateway access logs from stdin
fn main() {
    let mut counts: BTreeMap<String, u32> = BTreeMap::new();
    for key in ["ERROR", "INFO", "WARN"] {
        counts.insert(key.to_string(), 0);
    }
    let stdin = io::stdin();
    for line in stdin.lock().lines().map_while(Result::ok) {
        let mut it = line.split_whitespace();
        let (Some(ts), Some(level), Some(node)) = (it.next(), it.next(), it.next()) else {
            continue;
        };
        let msg = it.collect::<Vec<_>>().join(" ");
        if let Some(c) = counts.get_mut(level) {
            *c += 1;
            if level == "ERROR" {
                println!("\u{26a0} {ts} [{node}] {msg}");
            }
        }
    }
    println!("\n--- Summary ---");
    for (level, count) in &counts {
        println!("{level}: {count}");
    }
}
